Introduce apt repository config and deb resources

TeodorPt opened the pull request:
This PR introduces the possibility to optionally deploy from a custom apt repository, a custom TImescaleDB version, or from custom deb resource. By default, it will pull the latest version from the official third party repository.

Commit:
Linting

Commit:
Introduce optional deb resources

frankban commented on src/charm.py:
```diff
@@ -35,75 +34,101 @@ def _on_install(self, event):
+            debs = ["loader-deb", "tools-deb", "deb"]
+            deb_paths = []
+
+            for i in range(len(debs)):
```
Just `for deb in debs`?
And then `path = self.model.resources.fetch(deb)` ?
like `paths = [self.model.resources.fetch(deb) for deb in debs]`

frankban reviewed (approved):
Looks good!

Commit:
Fix GH Action Juju version

alesstimec reviewed (changes requested):
I don't think this approach will work quite as expected.. 

The install event is only emitted once per unit at the beginning of a charm's lifecycle (https://juju.is/docs/sdk/install-event). If at that point the user has not yet specified debs as resource, charm will install packages from the default repo. If you provide debs at a later stage by using "juju attach-resource" the install hook will not fire again and your debs will not be installed.

Also a more standard approach is for a user to provide a PPA and have packages installed from it instead of the standard repo.

TeodorPt commented:
> I don't think this approach will work quite as expected..
> 
> The install event is only emitted once per unit at the beginning of a charm's lifecycle (https://juju.is/docs/sdk/install-event). If at that point the user has not yet specified debs as resource, charm will install packages from the default repo. If you provide debs at a later stage by using "juju attach-resource" the install hook will not fire again and your debs will not be installed.
> 
> Also a more standard approach is for a user to provide a PPA and have packages installed from it instead of the standard repo.

Thanks for the input. Indeed, I did not consider adding resources post-install... I can add the custom PPA config for the charm, but I think debs could still be useful in case the charm is to be run in a very confined environment? Should we keep it around, with the caveat that it is a sub-par experience, or should we just remove it?

alesstimec commented:
that is up to you.. you can keep the debs, but make it clear in the readme, that it may not always work - you need to attach resource before the install hook fires..

alesstimec commented on .github/workflows/ci.yml:
```diff
@@ -19,6 +19,8 @@ jobs:
       - name: Setup operator environment
         uses: charmed-kubernetes/actions-operator@main
+        with:
+          juju-channel: 3.1/stable
```
oh, somebody likes to live on the edge ;)

Commit:
Move apt repository to config

TeodorPt commented:
I've run a test with the new code changes in the following way:
- deployed the app configured to TimescaleDB version 2.9.3
- tested that TimescaleDB works on a database by enabling the extension, verified the version
- changed the config to use latest
- observed the packages are updated
- tested that TimescaleDB still works and also triggered an update for the extension, verified the version is correct

Commit:
Fix integration test

brotich commented on config.yaml:
```diff
@@ -1 +1,21 @@
+      'apt-repository'. If it does not apply, configure to empty.
+    type: string
+  version:
+    default:
```
what does an empty default mean here?

TeodorPt replied:
It means the latest version is to be installed.

brotich reviewed (approved):
looks good with my limited charm knowledge

Commit:
Update on_upgrade hook

alesstimec reviewed (commented):
I would rather see the deb resources being handled in the update charm hook. Also instead of recording the "installed" flag in the state, i'd propose recording the hash of each of the deb package files provided. If the hash changes, you reinstall the package..

Commit:
Refactor helpers to not interact with event, remove unneeded config

This commit refactors all the helper functions to not receive the event or modify the unit status. They are instead agnostic of the events and throw exceptions
when they fail. The callers (i.e. hook functions) are the ones catching the exceptions and updating the unit status accordingly.

Also, removes unneeded config `from_resources` (it can be inferred from the presence or absence of resources at install time).

Also makes some small improvements, such as comparing hashes of installed debs when upgrading, for installations from resources.

Commit:
Add unit tests for charm

alesstimec commented on config.yaml:
```diff
@@ -1 +1,17 @@
+    default: https://packagecloud.io/timescale/timescaledb/gpgkey
+    description: |
+      The apt GPG key to use for installing TimescaleDB, in conjunction with
+      'apt-repository'. If it does not apply, configure to empty.
```
"configure to empty" does not apply, as it has a meaningful default value.

TeodorPt replied:
By "configure to empty" I meant "set the config to an empty-string value".

alesstimec reviewed (approved):
LGTM!
